### PR TITLE
Fix count/avg push down

### PR DESCRIPTION
**Pull request by shiyuhang0**
### What problem does this PR solve? 
close https://github.com/pingcap/tispark/issues/2313
 https://github.com/pingcap/tispark/pull/1840
close https://github.com/pingcap/tispark/issues/1142

### What is changed and how it works?
this bug is simply fixed by push down Constant(1) when count(1) or count(*)
 - push down count by add pattern match for count, but
    -  count(set) will not be pushed down because tikv can't accpect set type.
    - count(col1,col2,...) will not be pushed down because TiKV can't accpect (TIDB not support it)
 - push down Constant(1) when count(1) or count(*)
 - rewrite test and add test for count/avg

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/TiExprUtils.scala [hunk 1]
```diff
@@ -91,7 +98,7 @@ object TiExprUtils {
           }
 
           dagRequest.addRequiredColumn(firstColRef)
-          firstColRef
+          Constant.create(1, null)
```
Comment (xuanyu66) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/TiExprUtils.scala:101]: Add some code comments

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

**Comment by Daemonxiao** [comment]
Could you please add some content about how to solve this bug in "How it works"?

**Comment by shiyuhang0** [comment]
> Could you please add some content about how to solve this bug in "How it works"?

done

**Review by xuanyu66 — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/sql/pushdown/AvePushDownSuite.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.pushdown
+
+/**
+ * AVG only push down when count and sum can push down
+ * year type is not tested, because spark jdbc not support
+ */
+class AvePushDownSuite extends BasePushDownSuite() {
```
Comment (zhangyangyu) [inline core/src/test/scala/org/apache/spark/sql/pushdown/AvePushDownSuite.scala:23]: rename to avg plz

**Review by zhangyangyu — APPROVED** [review]

File: core/src/test/scala/org/apache/spark/sql/pushdown/AvgPushDownSuite.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    "select avg(tp_double) from ",
+    "select avg(tp_int) from ",
+    "select avg(tp_mediumint) from ",
+    "select avg(tp_real) from ",
+    "select avg(tp_smallint) from ",
+    "select avg(tp_tinyint) from ")
+
+  private val notPush = Seq[String]("select avg(tp_float) from ")
+
+  test("Test - Ave push down pk") {
```
Comment (zhangyangyu) [inline core/src/test/scala/org/apache/spark/sql/pushdown/AvgPushDownSuite.scala:38]:
```suggestion
  test("Test - Avg push down pk") {
```

**Pull request merged**